feat: add wordle proxy_program and target program.

aiqubits opened the pull request.

MedovTimur commented:
Hello,
Excellent work! I can see you put a lot of effort into this assignment. I have a question, and if you can answer it, I will grant you credit for the course.

aiqubits commented:
> Hello, Excellent work! I can see you put a lot of effort into this assignment. I have a question, and if you can answer it, I will grant you credit for the course.

Yeah, Looking forward to your question.

MedovTimur commented:
I asked a question in the review about `exec::wait_for(20)`

aiqubits commented:
> I asked a question in the review about `exec::wait_for(20)`

Compared to wait, wait_for will wait for a certain number of blocks without receiving a return message, causing the current thread to throw a panic. Here, wait_for(20) means waiting for 20 blocks, which is approximately 1 minute.

aiqubits commented:
> I asked a question in the review about `exec::wait_for(20)`

Compared to wait, wait_for wakes up the current thread after waiting for a certain number of blocks without receiving a return message. If payment still cannot be made after a certain number of blocks, the current thread throws a panic. Here wait_for(20) means waiting for 20 blocks, which is about 1 minute.

MedovTimur commented:
> > I asked a question in the review about `exec::wait_for(20)`
> 
> Compared to wait, wait_for wakes up the current thread after waiting for a certain number of blocks without receiving a return message. If payment still cannot be made after a certain number of blocks, the current thread throws a panic. Here wait_for(20) means waiting for 20 blocks, which is about 1 minute.

Okay, in your code, what happens after one minute?

aiqubits commented:
> > > I asked a question in the review about `exec::wait_for(20)`
> > 
> > 
> > Compared to wait, wait_for wakes up the current thread after waiting for a certain number of blocks without receiving a return message. If payment still cannot be made after a certain number of blocks, the current thread throws a panic. Here wait_for(20) means waiting for 20 blocks, which is about 1 minute.
> 
> Okay, in your code, what happens after one minute?

In this code, under normal circumstances, when the session contract sends a check word, it will wait for 1 minute. Within 1 minute, hand_deply receives a reply from the word contract, checks the hit status of the word, changes the Session Status, sends an event to the user, and then returns to the hand() function to match the branch based on the Session Status to determine whether to enter the next round of word check or end the game. If no response is received from the wordle contract within 1 minute, the current thread will be restored. Users can choose to try resending the check word to the wordle contract until they are unable to pay for the current message and throw a panic.

MedovTimur commented:
No, it works a bit differently, after a minute your message (word check) will start processing again (for the second time), but since you didn't change to some new status, it will go back to SessionStatus::Waiting and do the same thing again, it's better not to do it that way 
in the academy (https://academy.gear.foundation/courses/take/intermediate-course/multimedia/54008798-breakdowns-in-communication) you can see that before using wait() or wait_for() I changed the status to SessionStatus::MessageSent.
and this is done for a reason.

> > > > I asked a question in the review about `exec::wait_for(20)`
> > > 
> > > 
> > > Compared to wait, wait_for wakes up the current thread after waiting for a certain number of blocks without receiving a return message. If payment still cannot be made after a certain number of blocks, the current thread throws a panic. Here wait_for(20) means waiting for 20 blocks, which is about 1 minute.
> > 
> > 
> > Okay, in your code, what happens after one minute?
> 
> In this code, under normal circumstances, when the session contract sends a check word, it will wait for 1 minute. Within 1 minute, hand_deply receives a reply from the word contract, checks the hit status of the word, changes the Session Status, sends an event to the user, and then returns to the hand() function to match the branch based on the Session Status to determine whether to enter the next round of word check or end the game. If no response is received from the wordle contract within 1 minute, the current thread will be restored. Users can choose to try resending the check word to the wordle contract until they are unable to pay for the current message and throw a panic.

No, it works a bit differently, after a minute your message (word check) will start processing again (for the second time), but since you didn't change to some new status, it will go back to `SessionStatus::Waiting` and do the same thing again, it's better not to do it that way 
in the academy [link](https://academy.gear.foundation/courses/take/intermediate-course/multimedia/54008798-breakdowns-in-communication) you can see that before using `wait()` or `wait_for()` I changed the status to `SessionStatus::MessageSent`.
and this is done for a reason, but it is done so that if the response did not come, then when the same message is processed again, it already went to another status variant and reported that the response was not received.

aiqubits commented:
> No, it works a bit differently, after a minute your message (word check) will start processing again (for the second time), but since you didn't change to some new status, it will go back to SessionStatus::Waiting and do the same thing again, it's better not to do it that way 
> 
> in the academy (https://academy.gear.foundation/courses/take/intermediate-course/multimedia/54008798-breakdowns-in-communication) you can see that before using wait() or wait_for() I changed the status to SessionStatus::MessageSent.
> 
> and this is done for a reason.
> 
> 
> 
> > > > > I asked a question in the review about `exec::wait_for(20)`
> 
> > > > 
> 
> > > > 
> 
> > > > Compared to wait, wait_for wakes up the current thread after waiting for a certain number of blocks without receiving a return message. If payment still cannot be made after a certain number of blocks, the current thread throws a panic. Here wait_for(20) means waiting for 20 blocks, which is about 1 minute.
> 
> > > 
> 
> > > 
> 
> > > Okay, in your code, what happens after one minute?
> 
> > 
> 
> > In this code, under normal circumstances, when the session contract sends a check word, it will wait for 1 minute. Within 1 minute, hand_deply receives a reply from the word contract, checks the hit status of the word, changes the Session Status, sends an event to the user, and then returns to the hand() function to match the branch based on the Session Status to determine whether to enter the next round of word check or end the game. If no response is received from the wordle contract within 1 minute, the current thread will be restored. Users can choose to try resending the check word to the wordle contract until they are unable to pay for the current message and throw a panic.
> 
> 
> 
> No, it works a bit differently, after a minute your message (word check) will start processing again (for the second time), but since you didn't change to some new status, it will go back to `SessionStatus::Waiting` and do the same thing again, it's better not to do it that way 
> 
> in the academy [link](https://academy.gear.foundation/courses/take/intermediate-course/multimedia/54008798-breakdowns-in-communication) you can see that before using `wait()` or `wait_for()` I changed the status to `SessionStatus::MessageSent`.
> 
> and this is done for a reason, but it is done so that if the response did not come, then when the same message is processed again, it already went to another status variant and reported that the response was not received.

Yeah it's true. The code I wrote at the time didn't pay attention to the details here. If the wordle contract still doesn't respond, the session will continuously enter the check word branch, resend the check word, until a panic is thrown and the thread ends

aiqubits commented:
I will add a judgment to add a new state. Only when the word check is sent and no reply is received, will the user only reply with a sent event and exit the current thread
